quickwit/quickwit-indexing/src/source/queue_sources/coordinator.rs

```diff
@@ -0,0 +1,523 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+use std::sync::Arc;
+use std::time::Duration;
+
+use itertools::Itertools;
+use quickwit_actors::{ActorExitStatus, Mailbox};
+use quickwit_common::rate_limited_error;
+use quickwit_config::{FileSourceMessageType, FileSourceSqs};
+use quickwit_metastore::checkpoint::SourceCheckpoint;
+use quickwit_proto::indexing::IndexingPipelineId;
+use quickwit_proto::metastore::SourceType;
+use quickwit_storage::StorageResolver;
+use serde::Serialize;
+use tracing::info;
+use ulid::Ulid;
+
+use super::helpers::QueueReceiver;
+use super::local_state::QueueLocalState;
+use super::message::{MessageType, PreProcessingError, ReadyMessage};
+use super::shared_state::{checkpoint_messages, QueueSharedState};
+use super::visibility::{spawn_visibility_task, VisibilitySettings};
+use super::Queue;
+use crate::actors::DocProcessor;
+use crate::models::{NewPublishLock, NewPublishToken, PublishLock};
+use crate::source::{SourceContext, SourceRuntime};
+
+/// Maximum duration that the `emit_batches()` callback can wait for
+/// `queue.receive()` calls. If too small, the actor loop will spin
+/// un-necessarily. If too large, the actor loop will be slow to react to new
+/// messages (or shutdown).
+pub const RECEIVE_POLL_TIMEOUT: Duration = Duration::from_millis(500);
+
+#[derive(Default, Serialize)]
+pub struct QueueCoordinatorObservableState {
+    /// Number of bytes processed by the source.
+    pub num_bytes_processed: u64,
+    /// Number of lines processed by the source.
+    pub num_lines_processed: u64,
+    /// Number of messages processed by the source.
+    pub num_messages_processed: u64,
+    /// Number of messages that could not be pre-processed.
+    pub num_messages_failed_preprocessing: u64,
+    /// Number of messages that could not be moved to in-progress.
+    pub num_messages_failed_opening: u64,
+}
+
+pub struct QueueCoordinator {
```

Comment:
Can we have Docstring here explaining the basis about what the queue coordinator is for.